exporter/exporterhelper/internal/bounded_memory_queue_test.go

```diff
@@ -200,6 +211,28 @@ func TestZeroSize(t *testing.T) {
 	})
 
 	assert.False(t, q.Produce(newStringRequest("a"))) // in process
+
+	cs := newConsumerState(t)
+	q.OnOverflow(func(item Request) {
+		cs.record(item.(stringRequest).str)
+	})
+	assert.True(t, q.Produce(newStringRequest("b")))
+	cs.assertConsumed(map[string]bool{
+		"b": true,
+	})
+}
+
+func TestOverflow(t *testing.T) {
+	q := NewBoundedMemoryQueue(1)
+	cs := newConsumerState(t)
+	q.OnOverflow(func(item Request) {
+		cs.record(item.(stringRequest).str)
+	})
+	assert.True(t, q.Produce(newStringRequest("a")))
+	assert.True(t, q.Produce(newStringRequest("b")))
```

Comment:
What's the expectation of what happens to this request? Should the assert here be false as the queue should be overflowing?

Reply:
It overflows, but because we have an `OnOverflow` callback configured (which is only configured if there's a backlog), the head of the queue is removed and sent to it. So, the second request is successfully added to the queue and `Produce` returns true. In this test, we record the request that the overflow callback receives and assert that the correct one overflows.

```
// q.Produce(newStringRequest("a"))
capacity: 1
insert: "a"
before:
  queue: []
after:
  queue: ["a"]

// q.Produce(newStringRequest("b"))
capacity: 1
insert: "b"
before:
  queue: ["a"]
after:
  queue: ["b"]
  overflow: "a"
```